HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,72 @@
+#New version
+
+def check_and_spawn((key,(adj_list,weight, remove)), counter):
+	return_list = []
+	if weight == counter:
+		return_list.append((key,(adj_list,weight, 1)))    #set remove bit to 1 to remove it in the next iteration
+		for neighbor in adj_list:
+			return_list.append((neighbor,([], counter+1, -1)))
+	else:
+		return_list.append((key,(adj_list,weight, 0)))
+	return return_list
+
+def shrink_children((adj_list,weight, remove), (adj_list2,weight2, remove2) ,accum ):
+	if (len(list(adj_list)) == 0):
+		if weight2 == -1:
+			accum+=1
+			return (adj_list2, weight, remove2)
+		else:
+			return (adj_list2, weight2, remove2)
+	else:
+		if (weight == -1):
+			accum+=1
+			return (adj_list, weight2, remove)
+		else:
+			return (adj_list, weight, remove)
+
+
+def bfs(rdd, starting_node, sc):
+
+	#add a column for distance. initialize to 0 for starting node, and -1 for others
+	#add a second column (remove) for whether this node should be filtered out. ie. has it been reached from the starting node
+	bfs_rdd = rdd.map(lambda (x,y): (x,(y,0, 0)) if (x == starting_node) else (x,(y,-1, 0)) )
+
+	#result dictionary for the number of nodes touched in each iteration
+	results = {};
+	results[0] = 1
+	total_num = bfs_rdd.count() - 1
+
+	#intialize accumulator as 1 to start the while loop
+	accum = sc.accumulator(1)
+
+	counter = 0
+	while (accum.value != 0): 
+
+		#we only need to work with those records that hasn't been reached yet
+		filtered_rdd = bfs_rdd.filter(lambda (x,(y,z, remove)): remove == 0)
+		child_expanded_rdd = filtered_rdd.flatMap(lambda (x,(y,z, remove)): check_and_spawn( (x,(y,z, remove)), counter))
+		#print child_expanded_rdd.count()
+
+		#reset accumulator
+		accum = sc.accumulator(0)
```

Comment:
It would also be fine to check if your accumulator changed since the previous iteration